### PR TITLE
planner/core: Key partitioning uses an column index, which did not match value pairs in fast plan

### LINKED ISSUE #50206: Runtime error index out of range for Key partitioned table when partitioning column is not the first column.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
drop table if exists t;
create table t (b varchar(255), a int primary key nonclustered, key (b)) partition by key(a) partitions 3;
insert into t values ('Ab', 1),('abc',2),('BC',3),('AC',4),('BA',5),('cda',6);
analyze table t;
select * from t where a = 2;
```

### 2. What did you expect to see? (Required)
Success
### 3. What did you see instead (Required)
```sql
tidb> create table t (b varchar(255), a int primary key nonclustered, key (b)) partition by key(a) partitions 3;
Query OK, 0 rows affected (0.13 sec)

tidb> insert into t values ('Ab', 1),('abc',2),('BC',3),('AC',4),('BA',5),('cda',6);
Query OK, 6 rows affected (0.02 sec)
Records: 6  Duplicates: 0  Warnings: 0

tidb> analyze table t;
Query OK, 0 rows affected, 6 warnings (0.29 sec)

tidb> select * from t where a = 2;
ERROR 1105 (HY000): runtime error: index out of range [1] with length 1

```
### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v7.5.0
Edition: Community
Git Commit Hash: 069631e2ecfedc000ffb92c67207bea81380f020
Git Branch: heads/refs/tags/v7.5.0
UTC Build Time: 2023-11-24 08:50:14
GoVersion: go1.21.3
Race Enabled: false
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by bb7133**
This is an manual cherry-pick of #50603

### What problem does this PR solve?

Issue Number: close #50206

Problem Summary:
The column index was set to the second column, but in Fast Plan the value pairs is always the first one, so it needed to be adjusted for Key Partitioned tables.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Defined2014 — DISMISSED** [review]

**Review by tiancaiamao — APPROVED** [review]

File: pkg/table/tables/test/partition/partition_test.go [hunk 1]
```diff
@@ -3047,3 +3047,13 @@ func randStr(n int, r *rand.Rand) string {
...
+
+func TestPointGetKeyPartitioning(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec(`use test`)
+	tk.MustExec(`CREATE TABLE t (a VARCHAR(30) NOT NULL, b VARCHAR(45) NOT NULL,
+ c VARCHAR(45) NOT NULL, PRIMARY KEY (b, a)) PARTITION BY KEY(b) PARTITIONS 5`)
+	tk.MustExec(`INSERT INTO t VALUES ('Aa', 'Ab', 'Ac'), ('Ba', 'Bb', 'Bc')`)
+	tk.MustQuery(`SELECT * FROM t WHERE b = 'Ab'`).Check(testkit.Rows("Aa Ab Ac"))
+}
```
Comment (ywqzzy) [inline pkg/table/tables/test/partition/partition_test.go:3059]: Remove?

**Review by ywqzzy — APPROVED** [review]
LGTM

File: pkg/table/tables/test/partition/partition_test.go (hunk 1 above)
Comment (bb7133) [reply pkg/table/tables/test/partition/partition_test.go:3059]: I don't get it...why 'remove'?

Comment (ywqzzy) [reply pkg/table/tables/test/partition/partition_test.go:3059]: ut ft separation？

Comment (bb7133) [reply pkg/table/tables/test/partition/partition_test.go:3059]: Ah! I see, but this is a cherry-pick for hotfix, so that's fine.

**Review by Defined2014 — APPROVED** [review]

**Review by ywqzzy — APPROVED** [review]

**Pull request merged**